### ISSUE TITLE
"pm2 start" restarts already-running apps

### ISSUE THREAD

**Issue by c0bra** [Unitech/pm2#2577]
#### Expected behaviour
`pm2 start api` should not restart any instances of the `api` app if it's already running, per the docs: *By default, pm2 will only start a script if that script isn’t already running.*

#### Actual behaviour
The `api` app is stopped and started.

#### Steps to reproduce
1. Create a simple http server script
2. Start with pm2
3. Run `pm2 start <app name>` again

You will see the uptime reset to 0s and there will be stop/start of the app recorded in the pm2 logs.

#### Software versions used

```
OS         : OSX
node.js    : 6.7.0
PM2        : 2.2.2
```

#### PM2 Log output

```
2016-12-15 09:21:15: Stopping app:api id:4
2016-12-15 09:21:15: Stopping app:api id:6
2016-12-15 09:21:15: App [api] with id [4] and pid [1534], exited with code [0] via signal [SIGINT]
2016-12-15 09:21:15: App [api] with id [6] and pid [1533], exited with code [0] via signal [SIGINT]
2016-12-15 09:21:15: pid=1534 msg=process killed
2016-12-15 09:21:15: Starting execution sequence in -fork mode- for app name:api id:4
2016-12-15 09:21:15: pid=1533 msg=process killed
2016-12-15 09:21:15: Starting execution sequence in -fork mode- for app name:api id:6
2016-12-15 09:21:15: App name:api id:4 online
2016-12-15 09:21:15: App name:api id:6 online
2016-12-15 09:21:15: Stopping app:api id:8
2016-12-15 09:21:15: App [api] with id [8] and pid [1548], exited with code [0] via signal [SIGINT]
2016-12-15 09:21:15: pid=1548 msg=process killed
2016-12-15 09:21:15: Starting execution sequence in -fork mode- for app name:api id:8
2016-12-15 09:21:15: App name:api id:8 online
```

**Comment by vmarchaud** [comment]
Miss-understanding or our doc isn't clear but we don't start app that have the same name : 
`pm2 start api.js`: will have `api` as name
`pm2 start api.js`: will not be launched as it got the same name
`pm2 start api`: since you use the name, we understand that you want the already running application to restart

**Comment by c0bra** [comment]
@vmarchaud thanks for the clarification

I would say that yes, the docs are unclear. There's no mention (that I saw) that script name functions differently than app name.

As part of a deployment, if I wanted to make sure an app was already running without restarting it, I'd have to either hard-code the script path, parse the ecosystem config to get the path, or use the pm2 api to do the same thing.

This doesn't seem like desired behavior, does it?

**Comment by vmarchaud** [comment]
The file name of the script will be his name : if you start a script using `pm2 start path/to/my/script.js`, the name will be `script`.
If you try to re-run the command, pm2 will forbid it since the script is already started.
But if you use `pm2 start script`, pm2 will assume that you want to restart the application (since here `script` is the app name and not a path to a script).

If you want to do that, you can use `pm2 info <APP_NAME> && pm2 start myapp.js`, pm2 info will fail if the application doesn't exist so if the application exist, it doesn't execute the second command.

**Comment by c0bra** [comment]
I think `pm2 info <app> && pm2 start app.js` would actually only run `pm2 start` if the app already exists. `pm2 info` would succeed for an already-existing app.

You could do this hacky one-liner to start the app if it doesn't exist or is stopped:

`[[ $((pm2 info app || echo 'status|stopped') | grep status) =~ 'stopped' ]] && pm2 start app`

**Comment by Dzenly** [comment]
I can not force API function pm2.start to don't restart.
If I use the same options as for previous start, it stops previous processes.
If I use, just script name. It does not stop already running processes, but adds new processes to pm2 list.
If I use script and name - id stops previous processes.

So what is the way to don't stop already running processes but pm2.start() function ?

**Comment by vmarchaud** [comment]
@Dzenly the `force` option is only used when you start a specific file path (not just the script name like `index.js`, the full path of it)

**Comment by Dzenly** [comment]
The doc says that it is default behaviour to don't restart already running processes.
But the situation is vice-versa, it is hard as hell to don't stop already running processes.
BTW, how can I avoid such restarting using pm2.start ?

**Comment by vmarchaud** [comment]
My bad i misunderstood, could you make an issue a write clearly what you tried and what's the wanted behavior ?

**Comment by Dzenly** [comment]
I just whant to hear some way how to force pm2.start do not restart already running process.

You so clearly describe it for command line case.
```
pm2 start api.js: will have api as name
pm2 start api.js: will not be launched as it got the same name
pm2 start api: since you use the name, we understand that you want the already running application to restart
```
Now I wish the same description for pm2.start function :)

**Comment by MaxSvargal** [comment]
+1, that's behavior isn't correct and will been fixed probably?

**Comment by markmurphy242** [comment]
I just ran into this same issue.

Created a bash script test.sh that shows process id and sleeps for 10s.

Created a test.json config:

    instances       : 1,
    exec_mode       : "fork",
    interpreter     : "bash",
    autorestart     : false,
    merge_logs      : true

Called with:

pm2 start test.json

Expected that if 'pm2 start test.json' called while script is running, that original process is not reloaded and 2nd start has no effect.  Instead, first process is stopped and second process starts.

Tried:

pm2 start test.sh

All that did was change pm2's version of config with 'autorestart' set to true.  Yes, if 'pm2 start test.sh' is called while already running it states it will not reload **but** didn't expect it to change from autorestart:false to autorestart:true as I really just want the config test.json honored.

Expect a way to call pm2 that basically says if the process is already running as configured and a request to start again is issued, do not start.

Use case, running pm2 from a crontab and require single process running at a time.  If previous process still running cannot run again.

Basically a cheap pid lock without having to write one.  :)

**Comment by dhulke** [comment]
today I run some important scripts that cannot overlap in cron with flock. I thought I would be able to use pm2 for that, by creating an ecosystem file, and running pm2 start ecosyste.config.js --only job1. The problem is that that line stops the execution of the previous script to start the new one. Removing the job name and replacing it with the script name followed by a very long sequence of arguments is very nasty. Pm2 already has the `pm2 restart jobname`, why make start have almost the same behavior? We need a way to lock the execution of scripts by their job name

**Comment by majimboo** [comment]
I'm trying to start only the stopped services but it also restarts already running services which I don't want. How to get around this? 

    pm2 start all

**Comment by KevinBarbier** [comment]
I was having the same issues with using pm2 start/restart, running over pm2 start twice was showing weird behaviour with the app configured for 4 instances in cluster mode, pm2 left behind processes it didnt kill aswell as trying to restart the lingering process in fork mode eventhough the app is configured in cluster mode.

my fix was to basically delete the ecosystem file and then start it again, from package.json:

`pm2 del ./ecosystem/development/ecosystem.config.js && npm install && npm run build && pm2 start ./ecosystem/development/ecosystem.config.js --env development --update-env`

**Comment by vdegenne** [comment]
Here's how I would do it:
```bash
pm2 info <app-name> >/dev/null 2>&1 || pm2 start pm2.config.cjs
```
But I don't understand why after all this time of pm2 existing there is no a simple flag to do that, something like
```bash
pm2 start --prevent-restart pm2.config.cjs
```
👎

**Comment by John-H-Smith** [comment]
Sure this is working like expected?
I created a pm2config.json like that:
```
{
        "apps": [
                {
                        "name": "Microservice_1",
                        "script": "./microservice1/dist/main.js",
                        "instances": "1",
                        "exec_mode": "cluster",
                        "env": {
                            "NODE_ENV": "development"
                        },
                        "env_production" : {
                            "NODE_ENV": "production",
                            "PORT": 8080
                        },
                        "merge_logs": true,
                        "log_type": "json"
                },
               {
                        "name": "Microservice_2",
                        "script": "./microservice2/dist/main.js",
                        "instances": "1",
                        "exec_mode": "cluster",
                        "env": {
                            "NODE_ENV": "development"
                        },
                        "env_production" : {
                            "NODE_ENV": "production",
                            "PORT": 8081
                        },
                        "merge_logs": true,
                        "log_type": "json"
                }
        ]
}
```
When starting the app with `pm2 start pm2config.json`, I get the following output:
```
[PM2][WARN] Applications Microservice_1, Microservice_2 not running, starting...
[PM2] App [Microservice_1] launched (1 instances)
[PM2] App [Microservice_2] launched (1 instances)
┌────┬───────────────────┬─────────────┬─────────┬─────────┬──────────┬────────┬──────┬───────────┬──────────┬──────────┬──────────┬──────────┐
│ id │ name              │ namespace   │ version │ mode    │ pid      │ uptime │ ↺    │ status    │ cpu      │ mem      │ user     │ watching │
├────┼───────────────────┼─────────────┼─────────┼─────────┼──────────┼────────┼──────┼───────────┼──────────┼──────────┼──────────┼──────────┤
│ 0  │ Microservice_1    │ default     │ 0.0.1   │ cluster │ 3872950  │ 0s     │ 0    │ online    │ 0%       │ 0b       │ p672058  │ disabled │
│ 1  │ Microservice_2    │ default     │ 0.0.1   │ cluster │ 3872951  │ 0s     │ 0    │ online    │ 0%       │ 0b       │ p672058  │ disabled │
└────┴───────────────────┴─────────────┴─────────┴─────────┴──────────┴────────┴──────┴───────────┴──────────┴──────────┴──────────┴──────────┘
```

When running `pm2 start pm2config.json` a second time, I get:
```
[PM2] Applying action restartProcessId on app [Microservice_1](ids: [ 0 ])
[PM2] Applying action restartProcessId on app [Microservice_2](ids: [ 1 ])
[PM2] [Microservice_2](1) ✓
[PM2] [Microservice_1](0) ✓
┌────┬───────────────────┬─────────────┬─────────┬─────────┬──────────┬────────┬──────┬───────────┬──────────┬──────────┬──────────┬──────────┐
│ id │ name              │ namespace   │ version │ mode    │ pid      │ uptime │ ↺    │ status    │ cpu      │ mem      │ user     │ watching │
├────┼───────────────────┼─────────────┼─────────┼─────────┼──────────┼────────┼──────┼───────────┼──────────┼──────────┼──────────┼──────────┤
│ 0  │ Microservice_1    │ default     │ 0.0.1   │ cluster │ 3873050  │ 0s     │ 1    │ online    │ 0%       │ 0b       │ p672058  │ disabled │
│ 1  │ Microservice_2    │ default     │ 0.0.1   │ cluster │ 3873051  │ 0s     │ 1    │ online    │ 0%       │ 0b       │ p672058  │ disabled │
└────┴───────────────────┴─────────────┴─────────┴─────────┴──────────┴────────┴──────┴───────────┴──────────┴──────────┴──────────┴──────────┘
```
Thus, it seems like it is always restarting, even if giving a script as start option.
